### PR TITLE
Extend site-copy tool to include deep copy of files and folders using…

### PR DESCRIPTION
**Pull request by paulgconlon**
… RDM log projection. Design objects are copied directly into the design file service using multipart uploads. Design data, design sets and tasks are now reproduced at the destination site.

### CODE REVIEW

File: sites/site-tool/site-tool.py [hunk 1]
```diff
@@ -194,62 +255,207 @@ def hist_action(id):
...
+                        # assists with usability for web users and results in an RDM object that we don't want
+                        # to replicate at the destination site. Exclude it here based on the known UUID:
+                        if decoded_event["_id"] == "04585119-e2c2-4ed2-b336-5a30ca90c95f": # default_design_object_set
+                            continue
+
+                        copied, errors = copy_object_type(a_slug=slug, a_object_type=object_type, a_object_id="_id={}".format(decoded_event["_id"]), a_copy_count=copied, a_error_count=errors, a_decoded_event=decoded_event, a_rdm_accessor=dest_rdm_accessor)
+
+                    elif object_type == "sl::designObject" or object_type == "sl::deviceDesignObject":
+
+                        particular_dict = {
```
Comment (topcon-lrietberg) [inline sites/site-tool/site-tool.py:407]: please move these consts outside while loop.

File: sites/site-tool/site-tool.py [hunk 2]
```diff
@@ -194,62 +255,207 @@ def hist_action(id):
...
+
+                    elif object_type == "sl::list":
+                        ignored = ignore_object_type(a_slug=slug, a_object_type=object_type, a_object_id="_id={}".format(decoded_event["_id"]), a_ignore_count=ignored, a_additional_text="lists are special purpose objects")
+
+                    elif object_type == "sl::site":
+                        ignored = ignore_object_type(a_slug=slug, a_object_type=object_type, a_object_id="_id={}".format(decoded_event["_id"]), a_ignore_count=ignored, a_additional_text="the source site definition is not copied to the destination site")
+
+                    elif object_type == "sl::designObjectSet":
+
+                        # Design Object Sets look something like the following.
```
Comment (topcon-lrietberg) [inline sites/site-tool/site-tool.py:386]: would be good move designObject code to a separate function.

File: sites/site-tool/site-tool.py [hunk 3]
```diff
@@ -5,10 +5,66 @@
...
+    while True:
+        part = a_file_ptr.read(a_part_size)
+        if not part:
+            break
+        yield part 
+
+def upload_file_multipart(a_url, a_upload_uuid, a_file_location, a_file_name, a_media_type, a_encoding_type, a_jwt):
+    file_path = a_file_location + os.path.sep + a_file_name
+    part_index = 0
+    sitelink_max_part_size = 10485760
```
[the remark this answers is not in the document]
Comment (paulgconlon) [reply sites/site-tool/site-tool.py:29]: Refactored constant.

File: sites/site-tool/site-tool.py [hunk 4]
```diff
@@ -5,10 +5,66 @@
...
+        multipartEncoder = MultipartEncoder(fields=form_header)
+        headers = {
+            "Content-Type": "multipart/form-data; media-type=\"%s\"; boundary=%s" % (a_media_type, multipartEncoder.boundary_value),
+            "Authorization": "Bearer " + a_jwt,
+            "Content-Encoding": a_encoding_type
+        }
+        response = requests.post(a_url, headers=headers, data=multipartEncoder)
+        print("File part upload returned {}:{}".format(response.status_code, response.text))
+        if response.status_code != 200:
+            success = False
```
[the remark this answers is not in the document]
Comment (paulgconlon) [reply sites/site-tool/site-tool.py:64]: Broke loop. Found return statement was inside loop iteration too, so moved to outside such that return only happens after loop has finished iteration.

File: sites/site-tool/site-tool.py [hunk 5]
```diff
@@ -194,62 +255,207 @@ def hist_action(id):
...
+                        if upload_success:
+                            copied, errors = copy_object_type(a_slug=slug, a_object_type=object_type, a_object_id="uuid={}".format(decoded_event["uuid"]), a_copy_count=copied, a_error_count=errors, a_decoded_event=decoded_event, a_rdm_accessor=dest_rdm_accessor)
+                        else:
+                            print("Upload failed.")
+                            errors += 1
+
+                    elif object_type == "fs::folder":
+                        copied, errors = copy_object_type(a_slug=slug, a_object_type=object_type, a_object_id="_id={}".format(decoded_event["_id"]), a_copy_count=copied, a_error_count=errors, a_decoded_event=decoded_event, a_rdm_accessor=dest_rdm_accessor)
+
+            else:
```
[the remark this answers is not in the document]
Comment (paulgconlon) [reply sites/site-tool/site-tool.py:358]: Have explicitly handled the 204 we get on no more data and error returns separately.

File: sites/site-tool/site-tool.py [hunk 6]
```diff
@@ -194,62 +255,207 @@ def hist_action(id):
...
+        output_file_qualified_name = os.path.join(a_output_dir, a_output_file_name)
+        with open(output_file_qualified_name, "wb") as handle:
+            for data in tqdm(response.iter_content()):
+                handle.write(data)
+        return output_file_qualified_name
+
+    def copy_action(a_dest_url, a_dest_site, a_dest_token):
+        """ Use log projection to duplicate this RDM's log into a new site."""
+
+        dest_rdm_accessor, source_rdm_log_projection_url, params, source_headers = configure_log_projection(a_dest_url=a_dest_url, a_verify=not args.unverified, a_site=a_dest_site, a_domain="file_system", a_dest_token=a_dest_token)
```
[the remark this answers is not in the document]
Comment (paulgconlon) [reply sites/site-tool/site-tool.py:312]: Have broken this into a generic loop handler and individual callbacks for file_system and sitelink domain instances.

File: sites/site-tool/site-tool.py [hunk 7]
```diff
@@ -194,62 +255,207 @@ def hist_action(id):
...
+
+                        # Now upload the file under the same parent (if specified) at the dest site.
+                        url = "{}/file/v1/sites/{}/upload".format(a_dest_url, a_dest_site)
+                        upload_success = upload_file_multipart(a_url=url, a_upload_uuid=decoded_event["uuid"], a_file_location=output_dir, a_file_name=decoded_event["name"], a_media_type="multipart/mixed", a_encoding_type="binary", a_jwt=a_dest_token)
+
+                        if upload_success:
+                            copied, errors = copy_object_type(a_slug=slug, a_object_type=object_type, a_object_id="uuid={}".format(decoded_event["uuid"]), a_copy_count=copied, a_error_count=errors, a_decoded_event=decoded_event, a_rdm_accessor=dest_rdm_accessor)
+                        else:
+                            print("Upload failed.")
+                            errors += 1
```
[the remark this answers is not in the document]
Comment (paulgconlon) [reply sites/site-tool/site-tool.py:353]: Any errors will have been output by the time "Upload failed." is printed. This is done by post_rdm_payload call from within copy_object_type.

File: sites/site-tool/site-tool.py (hunk 1 above)
Comment (paulgconlon) [reply sites/site-tool/site-tool.py:407]: These have been refactored and now live inside event_callback_sitelink function.

File: sites/site-tool/site-tool.py (hunk 2 above)
Comment (paulgconlon) [reply sites/site-tool/site-tool.py:386]: This handling now happens in event_callback_sitelink. Is this sufficient?

File: sites/site-tool/site-tool.py [hunk 8]
```diff
@@ -194,62 +256,216 @@ def hist_action(id):
...
+        slug, decoded_event, object_type = process_log_event(a_event=a_event, a_status_dict=a_status_dict)
+
+        if object_type.startswith("_"):
+            ignore_object_type(a_slug=slug, a_object_type=object_type, a_object_id="_id={}".format(decoded_event["_id"]), a_status_dict=a_status_dict, a_additional_text="internal RDM type")
+
+        elif object_type == "fs::file":
+            # Download the file. Don't ignore archived files as they may be the origin of design file content.
+            get_file_url = "{}/file/v1/sites/{}/files/{}/url".format(args.url, args.site, decoded_event["uuid"])
+            res = requests.get(get_file_url, verify=not args.unverified, headers=a_source_headers)
+            while res.status_code == 503:
```
Comment (topcon-lrietberg) [inline sites/site-tool/site-tool.py:279]: do you mean statu code 504?

File: sites/site-tool/site-tool.py [hunk 9]
```diff
@@ -194,62 +256,216 @@ def hist_action(id):
...
+            # assists with usability for web users and results in an RDM object that we don't want
+            # to replicate at the destination site. Exclude it here based on the known UUID:
+            if decoded_event["_id"] == "04585119-e2c2-4ed2-b336-5a30ca90c95f": # default_design_object_set
+                return
+
+            copy_object_type(a_slug=slug, a_object_type=object_type, a_object_id="_id={}".format(decoded_event["_id"]), a_decoded_event=decoded_event, a_rdm_accessor=a_dest_rdm_accessor, a_status_dict=a_status_dict)
+
+        elif object_type == "sl::designObject" or object_type == "sl::deviceDesignObject":
+
+            particular_dict = {
```
Comment (topcon-lrietberg) [inline sites/site-tool/site-tool.py:339]: this const can be moved out of the loop.

File: sites/site-tool/site-tool.py [hunk 10]
```diff
@@ -194,62 +256,216 @@ def hist_action(id):
...
+            output_name = "{}.{}".format(decoded_event["name"],particular_dict[decoded_event["designType"]])
+            output_dir = os.path.join(os.path.dirname(os.path.realpath(__file__)), args.site)
+            print("{}: Processing design file from {}".format(slug, get_file_url))
+            download_file(a_source_url=get_file_url, a_source_headers=a_source_headers, a_output_file_name=output_name, a_output_dir=output_dir, a_source_site=args.site)
+
+            # Now upload that MAXML to the destination site.
+            upload_uuid = decoded_event["doFileUUID"]
+            design_file_uuid = decoded_event["doFileUUID"]
+
+            media_type_dict = {
```
Comment (topcon-lrietberg) [inline sites/site-tool/site-tool.py:358]: move const out of loop

File: sites/site-tool/site-tool.py [hunk 11]
```diff
@@ -194,62 +256,216 @@ def hist_action(id):
...
+            url = "{}/designfile/v1/sites/{}/design_files/{}/fineupload".format(a_dest_url, a_dest_site, design_file_uuid)
+            upload_success = upload_file_multipart(a_url=url, a_upload_uuid=design_file_uuid, a_file_location=output_dir, a_file_name=output_name, a_media_type=media_type, a_encoding_type="identity", a_jwt=a_dest_token)
+            
+            if upload_success:
+                # Ensure that the RDM payload contains a "createdAt" field.
+                if not "createdAt" in decoded_event:
+                    decoded_event["createdAt"] = decoded_event["_at"]
+                copy_object_type(a_slug=slug, a_object_type=object_type, a_object_id="_id={}".format(decoded_event["_id"]), a_decoded_event=decoded_event, a_rdm_accessor=a_dest_rdm_accessor, a_status_dict=a_status_dict)
+            else:
+                print("Upload failed.")
```
Comment (topcon-lrietberg) [inline sites/site-tool/site-tool.py:376]: add to error list

File: sites/site-tool/site-tool.py (hunk 9 above)
Comment (paulgconlon) [reply sites/site-tool/site-tool.py:339]: done.

File: sites/site-tool/site-tool.py (hunk 10 above)
Comment (paulgconlon) [reply sites/site-tool/site-tool.py:358]: done.

File: sites/site-tool/site-tool.py (hunk 11 above)
Comment (paulgconlon) [reply sites/site-tool/site-tool.py:376]: Great spot. We were doing this for fs::file but not sl::designObject

File: sites/site-tool/site-tool.py (hunk 8 above)
Comment (paulgconlon) [reply sites/site-tool/site-tool.py:279]: Will not fix as discussed on Slack.

**Review by topcon-lrietberg — APPROVED** [review]

**Pull request merged**